smart-py/marketplace.py

```diff
@@ -1,64 +1,203 @@
+            (params.objkt_id >= 152) &
+            (
+                (params.royalties >= 0) & (params.royalties <= 250)
+            )
+        )
+
+        # TODO
+        # i want to validate the royalties here to make sure they match
+        # what is set on the objkt when it is minted
+
+        # previous for easy comparison
+        # self.fa2_transfer(self.data.objkt, sp.sender, sp.self_address, params.objkt_id, params.objkt_amount)
+
+        self.fa2_transfer(
+            self.data.objkt,
+            sp.sender,
+            sp.self_address,
+            params.objkt_id,
+            params.objkt_amount
+        )
+
+        # previous for easy comparison
+        # self.data.swaps[self.data.counter] = sp.record(issuer=sp.sender, objkt_amount=params.objkt_amount, objkt_id=params.objkt_id, xtz_per_objkt=params.xtz_per_objkt, royalties=params.royalties, creator=params.creator)
+
+        self.data.swaps[self.data.counter] = sp.record(
+            issuer=sp.sender,
+            objkt_amount=params.objkt_amount,
+            objkt_id=params.objkt_id,
+            xtz_per_objkt=params.xtz_per_objkt,
+            royalties=params.royalties,
+            creator=params.creator
+        )
+
         self.data.counter += 1
-    
+
     @sp.entry_point
     def collect(self, params):
         sp.verify(
             # verifies if tez amount is equal to price per objkt
-            (sp.amount == sp.utils.nat_to_mutez(sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some()))) & (self.data.swaps[params.swap_id].objkt_amount != 0))
+            (
+                sp.amount == sp.utils.nat_to_mutez(
+                    sp.fst(
+                        sp.ediv(
+                            self.data.swaps[params.swap_id].xtz_per_objkt,
+                            sp.mutez(1)
+                        ).open_some()
+                    )
+                )
+            # and verify there are objkts available for swap
+            ) & (
+                self.data.swaps[params.swap_id].objkt_amount > 0
+            # and not collecting own swap
+            ) & (
+                sp.sender != self.data.swaps[params.swap_id].issuer
+            )
+        )
 
         sp.if (self.data.swaps[params.swap_id].xtz_per_objkt != sp.tez(0)):
+            self.amount = sp.fst(
+                sp.ediv(
+                    self.data.swaps[params.swap_id].xtz_per_objkt,
+                    sp.mutez(1)
+                ).open_some()
+            )
 
-            self.amount = sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some())
-                
             # calculate fees and royalties
             self.fee = self.amount * (self.data.swaps[params.swap_id].royalties + self.data.fee) / 1000
+
             self.royalties = self.data.swaps[params.swap_id].royalties * self.fee / (self.data.swaps[params.swap_id].royalties + self.data.fee)
-            
+
```

Comment:
i would suggest to refactor this code a bit:
self.fee = sp.utils.nat_to_mutez(self.amount * self.data.fee / 1000)
self.royalties = sp.utils.nat_to_mutez(self.amount * self.data.swaps[params.swap_id].royalties / 1000)
Those should be the amounts to be sent to the creator and the H=N wallet.
The amount to be sent to the wallet that swap the edition should be: sp.amount - self.fee - self.royalties

Reply:
when substituting either of the lines the tests throw this error:

```
[error] Type Error
sp.TNat is not sp.TMutez
```